### ISSUE TITLE
Using <wordexp.h> leads to "error: undefined symbol" / POSIX support?

### ISSUE THREAD

**Issue by sy2002** [emscripten-core/emscripten#10403]
I am a beginner to emcc and not sure, if emscripten is supposed to support `<wordexp.h>` at all (or in general any of the POSIX library's headers), so feel free to kill/close this issue if this is out of scope. (In this case I'll need to change the source I am porting right now to get rid of POSIX.)

Let's use the sample code from http://man7.org/linux/man-pages/man3/wordexp.3.html as an example:

```
#include <stdio.h>
#include <stdlib.h>
#include <wordexp.h>

int main(int argc, char **argv)
{
   wordexp_t p;
   char **w;
   int i;

   wordexp("[a-c]*.c", &p, 0);
   w = p.we_wordv;
   for (i = 0; i < p.we_wordc; i++)
       printf("%s\n", w[i]);
   wordfree(&p);
   exit(EXIT_SUCCESS);
}
```

Compiling this with any standard compiler works fine.

Using emcc this yields:

```
error: undefined symbol: wordexp
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
error: undefined symbol: wordfree
Error: Aborting compilation due to previous errors
```

**Comment by sbc100** [comment]
I looks like we were excluding it from the build.   I'll open a PR to enable it.   A lot of the functionality probably won't work (the sub-process stuff) but some of it might.

**Comment by sy2002** [comment]
@sbc100: Thank you! As emscripten is offering the virtual filesystem, wordexp will probably work pretty fine. I will test it and report back here, as soon as / if the PR will be accepted.

**Comment by shyam-lab** [comment]
Hi,

This issue is still persisting when I'm running the code on one of our c++ codebase, is there a resolution available?

**Comment by sbc100** [comment]
The problem is that the musl implementation of wordexp depends on fork and being able to run a shell as a subprocess, which under emscripten does not work.    See https://github.com/emscripten-core/emscripten/blob/17fd09c836bbe2de2a652b09b829060c3f2f5834/system/lib/libc/musl/src/misc/wordexp.c#L107-L122

Maybe some other implementation would work, but the musl one won't.

**Comment by sbc100** [comment]
FWIW that change to add wordexp.c was reverted in #10822

**Comment by stsp** [comment]
wordexp() is still missing.
Maybe this ticket should be re-opened?

**Comment by sbc100** [comment]
We don't currently have any plans to add workexp support, since the musl version won't work under emscripten.   If somebody wants to try to add an implementation that does work that would great.

**Comment by stsp** [comment]
>  If somebody wants to try to add an implementation that does work that would great.

No implementation is going to
fully work, because it supports
things like backticks to explicitly
do a shell command substitution.
But I use `WRDE_NOCMD` and I
hope others do too, as otherwise
this may be insecure.

Now there are the sloppy implementations
that call the shell right away. They check
`WRDE_NOCMD` only to filter out
the backticks, and then call the shell
anyway. That seems to include the musl
impl, bsd impl, libwordex impl.
But I've found the "full" version, i.e.
the one that does the minimum amount
of shell calls, and exactly 0 in case of the
use of `WRDE_NOCMD`:
https://cgit.uclibc-ng.org/cgi/cgit/uclibc-ng.git/tree/libc/misc/wordexp/wordexp.c

Would you like to adopt that version?
Otherwise you need to exclude the
`wordexp()` definition from headers.
In that case I'll be able to detect its
absence with autoconf, and use a
replacement myself.

**Comment by stsp** [comment]
I did a rip-off:
https://github.com/dosemu2/dosemu2/blob/devel/src/base/misc/wordexp.c
https://github.com/dosemu2/dosemu2/blob/devel/src/base/misc/wordexp.h
I disabled the shell execution and
argv indexing, while the rest is kept.
I made sure this compiles with emscripten.
So you can just import these 2 files.
I renamed funcs to `wordexp_lite()` and
`wordfree_lite()` so you'll have to rename
them back.

**Comment by sbc100** [comment]
Sadly I don't think we would be able to accept those files as they have a LGPL license.  I don't think we currently have any GPL or LGPL code in our core libraries.  Hopefully you can keep them in your tree where they should work fine? 

If other emscripten users find this useful we could perhaps find another non-GPL source for these, but I don't think there are many users of this function.

**Comment by stsp** [comment]
In this case I would suggest to
remove wordexp.h and allow people
to use autoconf-based fall-backs.

**Comment by stsp** [comment]
As currently only link fails, which is
not an adequate situation.

**Comment by sbc100** [comment]
Yes, that makes sense to me.   We already do that for several headers we don't support.

**Comment by stsp** [comment]
You may want to do the same with
ucontext.h and timer_xx() funcs.
I am getting link failure on those too.

**Comment by stsp** [comment]
And sigsuspend()

**Comment by stsp** [comment]
No, neither of the aforementioned
functions can be stubbed. They all
are important to have or replaced.
